src/libraries/System.Text.RegularExpressions/gen/UpgradeToRegexGeneratorAnalyzer.cs

```diff
@@ -0,0 +1,296 @@
+        private static void AnalyzeInvocation(OperationAnalysisContext context, INamedTypeSymbol regexTypeSymbol, HashSet<IMethodSymbol> staticMethodsToDetect)
+        {
+            // Ensure the invocation is a Regex static method.
+            IInvocationOperation invocationOperation = (IInvocationOperation)context.Operation;
+            IMethodSymbol method = invocationOperation.TargetMethod;
+            if (!method.IsStatic || !SymbolEqualityComparer.Default.Equals(method.ContainingType, regexTypeSymbol))
+            {
+                return;
+            }
+
+            // We need to save the parameters as properties so that we can save them onto the diagnostic so that the
+            // code fixer can later use that property bag to generate the code fix and emit the RegexGenerator attribute.
+            if (staticMethodsToDetect.Contains(method))
+            {
+                string? patternArgumentIndex = null;
+                string? optionsArgumentIndex = null;
+
+                // Validate that arguments pattern and options are constant and timeout was not passed in.
+                if (!TryValidateParametersAndExtractArgumentIndices(invocationOperation.Arguments, ref patternArgumentIndex, ref optionsArgumentIndex))
+                {
+                    return;
+                }
+
+                // Create the property bag.
+                ImmutableDictionary<string, string?> properties = ImmutableDictionary.CreateRange(new[]
+                {
+                    new KeyValuePair<string, string?>(PatternIndexName, patternArgumentIndex),
+                    new KeyValuePair<string, string?>(RegexOptionsIndexName, optionsArgumentIndex)
+                });
+
+                // Report the diagnostic.
+                SyntaxNode? syntaxNodeForDiagnostic = invocationOperation.Syntax;
+                Debug.Assert(syntaxNodeForDiagnostic != null);
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptors.UseRegexSourceGeneration, syntaxNodeForDiagnostic.GetLocation(), properties));
+            }
+        }
+
+        /// <summary>
+        /// Analyzes an object creation expression to see if the invocation is a call to one of the Regex constructors,
+        /// and checks if they could be using the source generator instead.
+        /// </summary>
+        /// <param name="context">The object creation context.</param>
+        private static void AnalyzeObjectCreation(OperationAnalysisContext context, INamedTypeSymbol regexTypeSymbol)
+        {
+            // Ensure the object creation is a call to the Regex constructor.
+            IObjectCreationOperation operation = (IObjectCreationOperation)context.Operation;
+            if (!SymbolEqualityComparer.Default.Equals(operation.Type, regexTypeSymbol))
+            {
+                return;
+            }
+
+            // If the constructor also has a timeout argument, then don't emit a diagnostic.
```

Comment:
I should have added more info into this comment, sorry about that. The main issue here is that Regex constructor and static methods take in a TimeSpan as a timeout argument, but since TimeSpan values are not accepted as attribute parameters, the RegexGeneratorAttribute take `int milleseconds` instead. I originally did have support and was doing the conversion from TimeSpan -> milliseconds on the fly, but @stephentoub pointed out that this was much more logic than we would ideally want to have in the analyzer, and it could be error prone as we would have to account for a lot of possible cases. For reference, here is what a very similar analyzer is doing to make this conversion so we would basically need to have something similar to this https://github.com/meziantou/Meziantou.Analyzer/blob/b6b41fa2a835fff4dc8cb2a58d692ea1a39f22ab/src/Meziantou.Analyzer/Rules/UseRegexSourceGeneratorAnalyzer.cs#L173-L313

Given it is very rare for folks to even use a timeout to begin with, we opted for now to not emit diagnostics when the call-site has a timeout parameter.